curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -0,0 +1,1092 @@
+                chunkPos = dcChunkPos + dcLen;
+            }
+        } else if ((chunkPos >= dcChunkPos) &&
+                   (chunkPos < dcChunkPos + dcLen)) {
+            char* cacheData = (*dcpIter)->GetData();
+            s3ClientAdaptor_->GetFsCacheManager()->Get(iter->second);
+            /*
+                     ----              ReadData
+                   ---------           DataCache
+            */
+            if (chunkPos + readLen <= dcChunkPos + dcLen) {
+                memcpy(dataBuf, cacheData + chunkPos - dcChunkPos, readLen);
+                readLen = 0;
+                break;
+                /*
+                         ----------              ReadData
+                       ---------                DataCache
+                */
+            } else {
+                memcpy(dataBuf, cacheData + chunkPos - dcChunkPos,
+                       dcChunkPos + dcLen - chunkPos);
+                readLen = chunkPos + readLen - dcChunkPos - dcLen;
+                dataBufOffset = dcChunkPos + dcLen - chunkPos + dataBufOffset;
+                chunkPos = dcChunkPos + dcLen;
+            }
+        } else {
+            continue;
+        }
+    }
+
+    if (readLen > 0) {
+        ReadRequest request;
+        request.index = index_;
+        request.len = readLen;
+        request.chunkPos = chunkPos;
+        request.bufOffset = dataBufOffset;
+        requests->emplace_back(request);
+    }
+    return;
+}
+
+DataCachePtr ChunkCacheManager::FindWriteableDataCache(
+    uint64_t chunkPos, uint64_t len,
+    std::vector<DataCachePtr>* mergeDataCacheVer) {
+    // todo performance optimization
+    WriteLockGuard writeLockGuard(rwLockWrite_);
+    auto iter = dataWCacheMap_.begin();
+    uint32_t i = 0;
+    for (; iter != dataWCacheMap_.end(); iter++) {
+        LOG(INFO) << "FindWriteableDataCache pos:" << chunkPos
+                  << ", len:" << len
+                  << ", datacache pos:" << iter->second->GetChunkPos()
+                  << ", datacache len:" << iter->second->GetLen();
+        if (((chunkPos + len) >= iter->second->GetChunkPos()) &&
+            (chunkPos <=
+             iter->second->GetChunkPos() + iter->second->GetLen())) {
+            DataCachePtr dataCache = iter->second;
+            std::vector<uint32_t> waitDel;
+            while (1) {
+                iter++;
+                i++;
+                if (iter == dataWCacheMap_.end()) {
+                    break;
+                }
+                if ((chunkPos + len) < iter->second->GetChunkPos()) {
+                    break;
+                }
+
+                mergeDataCacheVer->emplace_back(iter->second);
+                waitDel.push_back(i);
+            }
+
+            std::vector<uint32_t>::iterator iterDel = waitDel.begin();
+            for (; iterDel != waitDel.end(); iterDel++) {
+                auto iter = dataWCacheMap_.begin();
+                std::advance(iter, *iterDel);
+                dataWCacheMap_.erase(iter);
```

Comment:
if you erase here, then the index in waitDel is invalid now

Reply:
fixed